sfs/time/drivingfunction.py

```diff
@@ -251,7 +252,191 @@ def apply_delays(signal, delays):
-    for column, row in enumerate(delays_samples):
-        out[row:row + len(data), column] = data
-    return util.DelayedSignal(out, samplerate, offset_samples / samplerate)
+    out = np.zeros((delays_samples.max() + len(signal), len(delays_samples)))
+    for channel, cdelay in enumerate(delays_samples):
+        out[cdelay:cdelay + len(signal), channel] = signal
+    return out, offset_samples / fs
+
+    
+def nfchoa_25d_plane(x0, r0, npw, max_order=None, c=None, fs=44100, normalize=True):
+    """Vritual plane wave by 2.5-dimensional NFC-HOA.
+
+    Parameters
+    ----------
+    x0 : (N, 3) array_like
+        Sequence of secondary source positions.
+    r0 : float
+        Radius of the circular secondary source distribution
+    npw : (3,) array_like
+        Unit vector (propagation direction) of plane wave.
+    max_order : int
+        Ambisonics order
+    c : float, optional
+        Speed of sound
+
+    Returns
+    -------
+    delay : float
+        Overall delay (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : dictionary
+        Second-order section filters 
+    phaseshift : float
+        Phase shift
+
+    """
+    if max_order is None:
+        max_order =_max_order_circular_harmonics(len(x0), max_order)
+    if c is None:
+        c = defs.c
+
+    x0 = util.asarray_of_rows(x0)
+    npw = util.asarray_1d(npw)
+    phipw, _, _ = util.cart2sph(*npw)
+    phi0, _, _ = util.cart2sph(*x0.T)
+    
+    delay = -r0/c
+    weight = 2
+    sos = {}
+    for m in range(0, max_order+1):
+        _, p, _ = sig.besselap(m, norm='delay')
+        # TODO: modify "sig.besselap" for very high orders (>150)
+        s0 = np.zeros(m)
+        sinf = (c/r0)*p
+        z0 = np.exp(s0/fs)
+        zinf = np.exp(sinf/fs)
+        # TODO: select matched-z or bilinear transform
+        if normalize:
+            k = _normalize_gain(s0, sinf, z0, zinf, fs=fs)
+        else:
+            k = 1
+        sos[m] = sig.zpk2sos(z0, zinf, k, pairing='nearest')
+        # TODO: normalize the SOS filters individually?
+    phaseshift = phipw + np.pi - phi0
+    return delay, weight, sos, phaseshift
+
+
+def nfchoa_25d_point(x0, r0, xs, max_order=None, c=None, fs=44100, normalize=True):
+    """Point source by 2.5-dimensional NFC-HOA.
+
+    Parameters
+    ----------
+    x0 : (N, 3) array_like
+        Sequence of secondary source positions.
+    r0 : float
+        Radius of the circular secondary source distribution
+    xs : (3,) array_like
+        Virtual source position.
+    max_order : int
+        Ambisonics order
+    c : float, optional
+        Speed of sound
+
+    Returns
+    -------
+    delay : float
+        Overall delay (common to all secondary source)
+    weight : float
+        Overall weight (common to all secondary sources)
+    sos : dictionary
+        Second-order section filters 
+    phaseshift : float
+        Phase shift
+
+    """
```

Comment:
Add a reference to the help message. In sfs-matlab we used eq. (11) from Spors, Kuscher, Ahrens (2011).

Reply:
Thank you for the review.
The references are now added with the respective equation numbers.